crates/scheduler/lib.rs

```diff
@@ -0,0 +1,112 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink::prelude::vec::Vec;
+
+use ink::env::{DefaultEnvironment, Environment};
+use scale::{Decode, Encode};
+
+type Balance = <DefaultEnvironment as Environment>::Balance;
+type AccountId = <DefaultEnvironment as Environment>::AccountId;
+type BlockNumber = <DefaultEnvironment as Environment>::BlockNumber;
+
+pub struct SchedulerExtension;
+
+impl SchedulerExtension {
+    pub fn schedule(
+        origin: Origin,
+        when: BlockNumber,
+        maybe_periodic: Option<(BlockNumber, u32)>,
+        priority: u8,
+        call_input: ContractCallInput,
+    ) -> Result<(), SchedulerError> {
+        ::ink::env::chain_extension::ChainExtensionMethod::build(0x30001)
+            .input::<(
+                Origin,
+                BlockNumber,
+                Option<(BlockNumber, u32)>,
+                u8,
+                ContractCallInput,
+            )>()
+            .output::<Result<(), SchedulerError>, true>()
+            .handle_error_code::<SchedulerError>()
+            .call(&(origin, when, maybe_periodic, priority, call_input))
+    }
+
+    pub fn cancel(origin: Origin, when: BlockNumber, index: u32) -> Result<(), SchedulerError> {
+        ::ink::env::chain_extension::ChainExtensionMethod::build(0x30002)
+            .input::<(Origin, BlockNumber, u32)>()
+            .output::<Result<(), SchedulerError>, true>()
+            .handle_error_code::<SchedulerError>()
+            .call(&(origin, when, index))
+    }
+}
+
+#[derive(PartialEq, Eq, Copy, Clone, Encode, Decode, Debug)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum SchedulerError {
+    /// Failed to schedule a call
+    FailedToSchedule = 1,
+    /// Cannot find the scheduled call.
+    NotFound = 2,
+    /// Given target block number is in the past.
+    TargetBlockNumberInPast = 3,
+    /// Reschedule failed because it does not change scheduled time.
+    RescheduleNoChange = 4,
+    /// Attempt to use a non-named function on a named task.
+    Named = 5,
+    /// Origin Caller is not supported
+    OriginCannotBeCaller = 98,
+    /// Unknown error
+    RuntimeError = 99,
+    /// Unknow status code
+    UnknownStatusCode,
+    /// Encountered unexpected invalid SCALE encoding
+    InvalidScaleEncoding,
```

Comment:
No Because you get `InvalidScaleEncoding` on the ink! side:
```
impl From<scale::Error> for SchedulerError {
    fn from(_: scale::Error) -> Self {
        SchedulerError::InvalidScaleEncoding
    }
}
```
So no need to decode a Retval